### ISSUE TITLE
training: productionize RecurrentPPO training and recurrent-state evaluation

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#7847]
## Parent and Dependency

- Programme: #7845
- Blocked by: #7846
- Position: child B, local implementation and bounded smoke only.

## Archetype Metadata

```yaml
archetype: training-runtime-implementation
evidence_tier: implementation-smoke
resource: local
slurm_submission_authorized: false
```

## Objective

Turn the current minimal `scripts/training/train_recurrent_ppo.py` lane into a production-capable, config-first training and evaluation path for the exact matched contract frozen by #7846.

The implementation must support independent training seeds, recurrent-state-aware periodic evaluation, deterministic checkpoint selection, resumable artifacts, resource monitoring, and fail-closed provenance. It must not run the full comparison campaign or change its scientific design.

## Current Gap

The current runner establishes a true `sb3_contrib.RecurrentPPO` smoke path, but it is intentionally small. The existing implementation:

- constructs a model and calls `learn(...)`;
- writes one final model and a compact manifest;
- uses the first configured seed when constructing environments;
- does not yet provide the full periodic evaluation and best-checkpoint workflow used by mature PPO training lanes;
- does not yet prove recurrent-state handling through evaluation, vector environments, interruption, and resume;
- does not yet emit all artifact and monitoring fields required by #7845/#7846.

A list of five seeds in one YAML file is not evidence that five independent policies were trained.

## Scope

In scope:

- Productionizing the local `scripts/training/train_recurrent_ppo.py` lane against the frozen
  #7846 matched contract: independent training seeds, recurrent-state-aware periodic evaluation,
  deterministic checkpoint selection, resumable artifacts, resource monitoring, and fail-closed
  provenance.
- Bounded local smoke runs that prove the runtime contract end to end.

Out of scope:

- Running the full comparison campaign or changing its scientific design (#7845/#7846).
- SLURM submission (`slurm_submission_authorized: false`); this child is local-only.
- Any benchmark-result, ranking, superiority, causal, release, dissertation, or paper-facing claim.

## Inputs

- `scripts/training/train_recurrent_ppo.py`
- `scripts/training/train_ppo.py` and its existing environment/evaluation/checkpoint helpers, where reusable without hidden behavioral changes
- `configs/training/ppo/issue_4014_recurrent_ppo_lstm_smoke_matched.yaml`
- successor full-budget config and contract from #7846
- existing artifact-path, W&B, run-tracker, and model-registry conventions
- `sb3_contrib.RecurrentPPO` and `MultiInputLstmPolicy`

### Independent training seeds

- One training process, run identity, output directory, manifest, checkpoint lineage, and terminal status per training seed.
- A batch launcher may enumerate seeds, but every seed remains independently resumable and attributable.
- Seed identity must cover Python, NumPy, PyTorch, environment construction, scenario sampling, policy initialization, and evaluation where applicable.
- Duplicate run IDs or seed/output collisions fail before training.

### Recurrent state during training and evaluation

- Use the algorithm's native recurrent rollout handling during training.
- During all explicit evaluation, pass and update `lstm_states` and `episode_start` correctly.
- Reset recurrent state on `terminated`, `truncated`, environment reset, scenario replacement, robot respawn, and vector-environment index reassignment.
- Never share state between environment indices, policies, episodes, evaluation workers, or training seeds.
- Never restore an LSTM state from an unrelated checkpoint, episode, or partial rollout.
- Missing, malformed, non-finite, or shape-incompatible recurrent state fails closed.

### Evaluation and checkpoint selection

- Support deterministic periodic evaluation at the cadence frozen by #7846.
- Use the exact evaluation scenario/seed manifests and checkpoint-selection metric from #7846.
- Record every evaluated checkpoint, not only the selected checkpoint.
- Keep the selected-checkpoint rule fixed before outcomes.
- Save at least: latest recoverable checkpoint, each scheduled evaluation checkpoint or immutable pointer, selected best checkpoint, and final checkpoint.
- A failed evaluation cannot silently leave an older checkpoint labelled as current-best without an explicit status.

### Resume behavior

- Resume only from a checkpoint whose source SHA, config digest, observation/action contract, algorithm version, policy class, seed, and environment identity match.
- Record previous and resumed environment-step counts and reject overlap or step regression.
- Do not resume recurrent rollout state across process restarts unless the exact rollout-state contract is implemented, versioned, and tested. Restart from a clean episode boundary by default.
- Preserve prior logs and checkpoint indexes; do not overwrite terminal evidence.

## Implementation Guidelines

1. Keep `train_recurrent_ppo.py` as the canonical recurrent entry point unless #7846 selects another existing owner.
2. Reuse stable PPO environment and artifact helpers, but do not bind recurrent training to private helper behavior without characterization tests.
3. Keep configuration strict. Reject unknown recurrent hyperparameters and policy kwargs.
4. Preserve `default_gym` non-nested observation compatibility for the first programme stage.
5. Keep the model output and environment action contract identical to #7846. Do not add a second action adapter here.
6. Use atomic writes for manifests, indexes, and compact monitoring output.
7. Keep local CPU smoke budgets small and explicit. Do not make a local smoke look like campaign evidence.
8. Keep optional W&B logging non-authoritative. Local and durable manifests remain sufficient to reconstruct the run.
9. Record package versions for Python, NumPy, PyTorch, Gymnasium, Stable-Baselines3, SB3-Contrib, Robot SF, CUDA, and relevant drivers when available.
10. Do not change feed-forward PPO defaults, benchmark metrics, safety-wrapper behavior, or model-release policy.

## Required Monitoring

Emit machine-readable monitoring with a stable schema. At minimum:

### Optimization

- environment steps and update count;
- policy-gradient loss;
- value loss;
- entropy loss or policy entropy;
- approximate KL divergence;
- clip fraction and clip range;
- explained variance;
- learning rate;
- gradient norm when exposed safely;
- non-finite loss/parameter/gradient checks.

### Runtime and resources

- elapsed wall time and training steps/s;
- evaluation wall time;
- CPU utilization and resident memory;
- GPU device, utilization, allocated/reserved/peak memory where available;
- worker count and vector-environment mode;
- checkpoint save duration and artifact size;
- terminal process and scheduler state when supplied by an outer launcher.

### Policy behavior

- action mean, standard deviation, extrema, and non-finite count;
- action-space clipping/saturation count;
- realized speed and acceleration summaries where available;
- termination and truncation reason counts;
- progress, low-progress, timeout, success, collision, near-miss, and clearance summaries from periodic evaluation;
- median/p95 inference latency during evaluation.

### Recurrent integrity

- recurrent-state shape and policy-layer metadata;
- state reset count grouped by reason;
- per-environment state ownership diagnostics;
- non-finite hidden/cell-state count;
- unexpected state carry-over count;
- maximum and distribution summary of hidden/cell-state norms as diagnostic data, without treating magnitude alone as quality evidence.

## Stop and Alert Conditions

Stop the affected training seed and preserve a terminal packet on:

- non-finite policy parameters, losses, gradients, actions, or recurrent state;
- observation/action shape drift;
- recurrent state assigned to the wrong environment index or episode;
- source/config/checkpoint digest mismatch;
- environment-step regression or duplicate checkpoint identity;
- persistent zero-motion or constant-saturated action behavior under a predeclared diagnostic rule;
- GPU/host out-of-memory, repeated worker crash, or corrupted checkpoint;
- missing required evaluation rows or artifact destination.

Resource or infrastructure failures may be retryable only after a changed environment or packet is recorded. Do not automatically repeat an identical scientific run fingerprint.

## Artifact Contract

Each seed must produce or reference:

```text
resolved_config.yaml
training_manifest.json
environment_contract.json
observation_contract.json
action_contract.json
dependency_environment.json
training_metrics.jsonl
evaluation_history.jsonl
checkpoint_index.json
latest.zip or immutable pointer
best.zip or immutable pointer
final.zip or immutable pointer
SHA256SUMS
```

Large checkpoints and logs remain in approved durable storage. Git may contain only compact manifests, checksums, and review evidence.

## Suggested Paths

- `scripts/training/train_recurrent_ppo.py`
- the smallest reusable helper under `robot_sf/training/`
- successor configs under `configs/training/ppo/`
- focused tests under `tests/training/`
- concise runtime documentation under `docs/context/`

Do not modify benchmark metric owners, safety-wrapper contracts, release manifests, or publication registries in this child.

## Acceptance Criteria

- [ ] A single-seed dry run validates all contracts without importing unavailable optional runtime services.
- [ ] A bounded non-dry CPU smoke trains, evaluates recurrently, writes checkpoints, and terminates cleanly.
- [ ] Two explicitly requested seeds create separate run identities, outputs, model initializations, and manifests.
- [ ] Periodic evaluation propagates recurrent state and `episode_start` correctly.
- [ ] Vector-environment state remains isolated by index and resets on every declared boundary.
- [ ] Resume succeeds from one compatible clean episode-boundary checkpoint and rejects mismatched source/config/seed/space identities.
- [ ] Selected-checkpoint output is reproducible from the evaluation history and frozen rule.
- [ ] Non-finite, shape-drift, state-leakage, duplicate-output, and corrupted-checkpoint fixtures fail closed.
- [ ] Monitoring and artifact schemas contain all required fields or explicit unavailable statuses.
- [ ] No full-budget training, SLURM submission, benchmark promotion, or model publication occurs.

## Validation

Add exact commands during implementation. Minimum expected coverage:

```bash
uv run pytest -q tests/training -k "recurrent_ppo and (config or state or eval or resume or checkpoint or monitoring)"

uv run python scripts/training/train_recurrent_ppo.py \
  --config <successor-smoke-config> --dry-run

uv run python scripts/training/train_recurrent_ppo.py \
  --config <successor-smoke-config> \
  --run-id <bounded-local-smoke>

uv run ruff check scripts/training/train_recurrent_ppo.py robot_sf/training tests/training
uv run ruff format --check scripts/training/train_recurrent_ppo.py robot_sf/training tests/training
git diff --check
```

Run final repository readiness for the exact implementation head.

## Stop Conditions

- Stop if productionization would require changing the #7846 scientific contract.
- Stop if a broad `train_ppo.py` refactor is required; extract a separately characterized dependency.
- Stop if correct recurrent evaluation cannot be implemented through the current SB3-Contrib API.
- Stop before SLURM submission or full training.

## Handoff

On completion, post to #7845 and the recurrence-campaign child:

- exact merged SHA;
- supported config path;
- smoke command and result;
- recurrent-state and resume guarantees;
- monitoring schema/version;
- unresolved runtime limits;
- artifact and checkpoint contract.

## Claim Boundary

Completion proves training-runtime and recurrent-evaluation integrity at smoke tier. It does not prove policy quality, safety, sample efficiency, or superiority over feed-forward PPO.
